src/components/Grid/Grid.tsx

```diff
@@ -0,0 +1,725 @@
+          target.dataset.gridColumn === undefined
+        ) {
+          return;
+        }
+        if (e.buttons === RIGHT_BUTTON_PRESSED && target.dataset.selected === "true") {
+          return;
+        }
+
+        const { gridRow: rowIndexString, gridColumn: columnIndexString } = target.dataset;
+        const row = parseInt(rowIndexString);
+        const column = parseInt(columnIndexString);
+        if (row === -1 && column === -1) {
+          onSelection({ type: "all" });
+          return;
+        }
+        const shiftKeyPressed = e.shiftKey;
+
+        if (row === -1) {
+          onSelection({
+            type: shiftKeyPressed ? "shiftColumnSelection" : "columnSelection",
+            column,
+            event: "click",
+          });
+          if (!shiftKeyPressed) {
+            onFocusChange(0, column);
+          }
+          return;
+        }
+        if (column === -1) {
+          onSelection({
+            type: shiftKeyPressed ? "shiftRowSelection" : "rowSelection",
+            row,
+            event: "click",
+          });
+          if (!shiftKeyPressed) {
+            onFocusChange(row, 0);
+          }
+          return;
+        }
+        onSelection({
+          type: shiftKeyPressed ? "shiftSelection" : "normal",
+          row,
+          column,
+          event: "click",
+        });
+        if (!shiftKeyPressed) {
+          onFocusChange(row, column);
+        }
+      },
+      [onFocusChange, onSelection]
+    );
+
+    const onPointerDown: PointerEventHandler<HTMLDivElement> = useCallback(e => {
+      dragState.current = e.pointerId;
+    }, []);
+
+    const setPointerCapture: PointerEventHandler<HTMLDivElement> = useCallback(e => {
+      if (e.buttons === NO_BUTTONS_PRESSED || dragState.current === false) {
+        return;
+      }
+
+      containerRef.current?.setPointerCapture(dragState.current);
+    }, []);
+
+    const onPointerUp: PointerEventHandler<HTMLDivElement> = useCallback(
+      e => {
+        if (containerRef.current) {
+          containerRef.current.releasePointerCapture(e.pointerId);
+          if (!contextMenuOpen) {
+            containerRef.current.focus();
+          }
+        }
+
+        dragState.current = false;
+        onFocusChange(focus.row, focus.column);
+      },
+      [contextMenuOpen, focus.column, focus.row, onFocusChange]
+    );
+
+    const onMouseMove: MouseEventHandler<HTMLDivElement> = useCallback(
+      async e => {
+        e.preventDefault();
+        e.stopPropagation();
+        if (
+          dragState.current === false ||
+          e.buttons === NO_BUTTONS_PRESSED ||
+          !containerRef.current ||
+          !outerRef.current
+        ) {
+          return;
+        }
+
+        containerRef.current.setPointerCapture(dragState.current);
+        let verticalSign = 0,
+          horizontalSign = 0,
+          x = e.clientX,
+          y = e.clientY;
+        if (
+          elementBorderRef.current.left >= e.clientX ||
```

Comment:
we need to add comments to these conditions, otherwise in 6months no one will know what it is doing.